extra/dist/tarantoolctl.in

```diff
@@ -506,6 +508,90 @@ local function start_check()
     return pid
 end
 
+ffi.cdef([[
+    pid_t fork(void);
+    int execve(const char *pathname, char *const argv[], char *const envp[]);
+    int dup2(int oldfd, int newfd);
+    int fileno(struct FILE *stream);
+]])
+
+local function ffi_table_to_const_char(input)
+    local result = ffi.new('char const*[?]', #input + 1, input)
+    result[#input] = nil
+    return ffi.cast('char *const*', result)
+end
+
+-- Starts process and returns immediately, not waiting until process is finished.
+-- @param path Executable path.
+-- @param[opt] args
+-- @param[opt] env
+local function fork_execve(path, args, env)
+    args = args or {}
+    env = env or {}
+    table.insert(args, 1, path)
+    local argv = ffi_table_to_const_char(args)
+    local env_list = fun.iter(env):map(function(k, v) return k .. '=' .. v end):totable()
+    local envp = ffi_table_to_const_char(env_list)
+    local pid = tonumber(ffi.C.fork())
+    if pid == -1 then
+        error('fork failed: ' .. errno.strerror())
+    elseif pid > 0 then
+        return pid
+    end
+    ffi.C.execve(path, argv, envp)
+    io.stderr:write('execve failed: ' .. errno.strerror() .. '\n')
+    os.exit(1)
+end
+
+-- Create socket and run forked tarantool with NOTIFY_SOCKET.
+--
+-- `box.cfg` may not be called inside init script and tarantool
+-- will not be daemonized (forcing `background = true` in hijacked `box.cfg`),
+-- so we need to fork tarantool before running this script.
+--
+-- To find out the moment when tarantool finishes startup it waits for `READY=1`
+-- message on NOTIFY_SOCKET which is emitted after init script is finished.
+local function fork_and_start_over()
+    local sock = assert(socket('AF_UNIX', 'SOCK_DGRAM', 0), 'Can not create socket')
+    local sock_name = '/tmp/tntctl-' .. require('uuid').str()
+    local ok = sock:bind('unix/', sock_name)
+    assert(ok, sock:error())
+    fio.chown(sock_name, default_cfg.username, group_name)
+    fio.chmod(sock_name, tonumber('0770', 8))
+
+    local env = table.copy(os.environ())
+    env.NOTIFY_SOCKET = sock_name
+    env.TARANTOOLCTL_DAEMONIZED = 'true'
+    -- slice 1..#arg indices into array-table
+    local args = fun.iter(arg):totable()
+    table.insert(args, 1, arg[0])
+    local pid = fork_execve(arg[-1], args, env)
+
+    local function exit(code)
+        fio.unlink(sock_name)
+        os.exit(code)
+    end
+
+    local function check_process_alive()
+        if os.execute('ps -p ' .. pid .. ' > /dev/null') ~= 0 then
```

Comment:
Why you don't use `waitpid()` here?

Reply:
I see 3 `waitpid`'s signatures and they look quite complicated for using with ffi. Also it looks like there should be separate thread for `waitpid` because it's blocking call.

Reply:
from the same man page:

```
The waitpid() function is provided for three reasons:

1. To support job control
2. To permit a non-blocking version of the wait() function
3. To permit a library routine, such as system() or pclose(), to wait for its children without interfering with other terminated children for which the process has not waited
```

See 2nd point

Reply:
It requires WNOHANG flag to be non-blocking. Hardcoding int value into ffi call can be not-portable.

Reply:
Once popen is merged in master, this code can be changed to use that features instead.